rfd/0144-client-tools-updates.md

```diff
@@ -0,0 +1,295 @@
+able to watch for changes to client tools versions which can then be used to
+trigger other integrations (using MDM software like JamF) to update the
+installed version of client tools on endpoints.
+
+```
+$ tctl autoupdate watch
+{"tools_version": "1.0.0"}
+{"tools_version": "1.0.1"}
+{"tools_version": "2.0.0"}
+
+[...]
+```
+
+##### Cluster configuration
+
+Enrollment of clients in automatic updates will be enforced at the cluster
+level.
+
+The `cluster_maintenance_config` resource will be updated to allow cluster
+administrators to turn client tools automatic updates `on` (default) or `off`.
+A `autoupdate_version` resource will be added to allow cluster administrators
+to manage the version of tools pushed to clients.
+
+> [!NOTE]
+> Client tools configuration is broken into two resources to [prevent
+> updates](https://github.com/gravitational/teleport/blob/master/lib/modules/modules.go#L332-L355)
+> to `autoupdate_verson` on Cloud.
+>
+> While Cloud customers will be able to use `cluster_maintenance_config` to
+> turn client tools automatic updates `off` and self-manage updates, they will
+> not be able to control the version of client tools in `autoupdate_verson`.
+> That will continue to be managed by the Teleport Cloud team.
+
+Both resources can either be updated directly or by using `tctl` helper
+functions.
+
+```yaml
+kind: cluster_maintenance_config
+spec:
+  # tools_auto_update allows turning client tools updates on or off at the
+  # cluster level. Only turn client tools automatic updates off if self-managed
+  # updates are in place.
+  tools_auto_update: on|off
+
+  [...]
+```
+```
+$ tctl autoupdate update --set-tools-auto-update=off
+Automatic updates configuration has been updated.
+```
+
+All Cloud clusters will be opted into `tools_auto_update: on`.
+
+All new self-hosted clusters will be opted into `tools_auto_update: on`.
+Existing self-hosted clusters will be opted into `tools_auto_update: off` to
+not introduce unexpected behavior.
+
+```yaml
+kind: autoupdate_version
+spec:
+  # tools_version is the version of client tools the cluster will advertise.
+  # Can be auto (match the version of the proxy) or an exact semver formatted
+  # version.
+  tools_version: auto|X.Y.Z
+```
+```
+$ tctl autoupdate update --set-tools-version=1.0.1
+Automatic updates configuration has been updated.
+```
+
+For Cloud clusters, `tools_version` will always be `X.Y.Z`, with the version
+controlled by the Cloud team.
+
+For self-hosted clusters, the default will be `tools_version: auto` matching
+the tools version with the proxy version. This way, automatic updates will be
+available to self-hosted customers by default, without any extra effort from
+the cluster administrator.
+
+The above configuration will then be available from the unauthenticated
+endpoint `/v1/webapi/ping` which clients will consult.
+
+```
+curl https://proxy.example.com/v1/webapi/ping | jq .
+{
+    "tools_auto_update": true,
```

Comment:
For agents, I'm recommending we include the following fields:
- Version
- Upgrade time OR immediate (can be a specific timestamp -- Cloud can translate window -> time)
- Jitter duration

Worth including for client tools as well, so we can stage rollouts of new client code?

https://github.com/gravitational/teleport/pull/39217/files#r1546654653

Reply:
No objections to adding it to `/v1/webapi/automaticupdates`.